test/suites/scale/deprovisioning_test.go

```diff
@@ -0,0 +1,630 @@
+)
+
+const (
+	deprovisioningTypeKey = v1alpha5.TestingGroup + "/deprovisioning-type"
+	consolidationValue    = "consolidation"
+	emptinessValue        = "emptiness"
+	expirationValue       = "expiration"
+	noExpirationValue     = "noExpiration"
+	driftValue            = "drift"
+	noDriftValue          = "noDrift"
+)
+
+const (
+	multipleDeprovisionersTestGroup = "multipleDeprovsioners"
+	consolidationTestGroup          = "consolidation"
+	emptinessTestGroup              = "emptiness"
+	expirationTestGroup             = "expiration"
+	driftTestGroup                  = "drift"
+	interruptionTestGroup           = "interruption"
+
+	defaultTestName = "default"
+)
+
+// disableProvisioningLimits represents limits that can be applied to a provisioner if you want a provisioner
+// that can deprovision nodes but cannot provision nodes
+var disableProvisioningLimits = &v1alpha5.Limits{
+	Resources: v1.ResourceList{
+		v1.ResourceCPU:    resource.MustParse("0"),
+		v1.ResourceMemory: resource.MustParse("0Gi"),
+	},
+}
+
+var _ = Describe("Deprovisioning", Label(debug.NoWatch), Label(debug.NoEvents), func() {
+	var provisioner *v1alpha5.Provisioner
+	var provisionerOptions test.ProvisionerOptions
+	var nodeTemplate *v1alpha1.AWSNodeTemplate
+	var deployment *appsv1.Deployment
+	var deploymentOptions test.DeploymentOptions
+	var selector labels.Selector
+	var dsCount int
+
+	BeforeEach(func() {
+		env.ExpectSettingsOverridden(map[string]string{
+			"featureGates.driftEnabled": "true",
+		})
+		nodeTemplate = awstest.AWSNodeTemplate(v1alpha1.AWSNodeTemplateSpec{AWS: v1alpha1.AWS{
+			SecurityGroupSelector: map[string]string{"karpenter.sh/discovery": settings.FromContext(env.Context).ClusterName},
+			SubnetSelector:        map[string]string{"karpenter.sh/discovery": settings.FromContext(env.Context).ClusterName},
+		}})
+		provisionerOptions = test.ProvisionerOptions{
+			ProviderRef: &v1alpha5.MachineTemplateRef{
+				Name: nodeTemplate.Name,
+			},
+			Requirements: []v1.NodeSelectorRequirement{
+				{
+					Key:      v1alpha1.LabelInstanceSize,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{"4xlarge"},
+				},
+				{
+					Key:      v1alpha5.LabelCapacityType,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{v1alpha1.CapacityTypeOnDemand},
+				},
+				{
+					Key:      v1.LabelOSStable,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{string(v1.Linux)},
+				},
+			},
+			// No limits!!!
+			// https://tenor.com/view/chaos-gif-22919457
+			Limits: v1.ResourceList{},
+		}
+		provisioner = test.Provisioner(provisionerOptions)
+		deploymentOptions = test.DeploymentOptions{
+			PodOptions: test.PodOptions{
+				ResourceRequirements: v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceCPU:    resource.MustParse("100m"),
+						v1.ResourceMemory: resource.MustParse("100Mi"),
+					},
+				},
+				TerminationGracePeriodSeconds: lo.ToPtr[int64](0),
```

Comment:
Just a thought, I think this is fine for now, but we should really start testing with some non-pause pods and some non-deterministic pod shutdown times. We're taking the assumption that pod graceful shutdown goes well here, which is definitely a place where deprovisioning can be bottlenecked.

Reply:
Agreed. Makes sense. We can do this as we start to expand the world of what we do in scale testing.